crates/rust-analyzer/src/line_index.rs

```diff
@@ -19,6 +22,66 @@ pub(crate) struct LineIndex {
     pub(crate) index: Arc<ide::LineIndex>,
     pub(crate) endings: LineEndings,
     pub(crate) encoding: PositionEncoding,
+    pub(crate) transform: PositionTransform,
+}
+
+impl LineIndex {
+    pub(crate) fn line_col(&self, mut offset: TextSize) -> ide::LineCol {
+        if !self.transform.insertions.is_empty() {
+            offset += TextSize::new(
+                self.transform
+                    .insertions
+                    .iter()
+                    .copied()
+                    .take_while(|&(off, _)| off < offset)
+                    .map(|(_, len)| ws_kind_width(len))
+                    .sum::<u32>(),
+            );
+        }
+        self.index.line_col(offset)
+    }
+
+    pub(crate) fn line(&self, line: u32) -> Option<ide::TextRange> {
+        let mut range = self.index.line(line)?;
+        if !self.transform.insertions.is_empty() {
+            let mut iter = self.transform.insertions.iter();
+            let mut nl_seen = 0;
+            let overall_sub = TextSize::new(
+                iter.peeking_take_while(|&&(_, ws)| {
+                    let m = nl_seen != line;
+                    if ws == PrettifyWsKind::Newline {
+                        nl_seen += 1;
+                    }
+                    m
+                })
+                .copied()
+                .map(|(_, len)| ws_kind_width(len))
+                .sum::<u32>(),
+            );
+            let end_sub = TextSize::new(
+                iter.copied()
+                    .take_while(|&(_, ws)| ws != PrettifyWsKind::Newline)
+                    .map(|(_, len)| ws_kind_width(len))
+                    .sum::<u32>(),
+            );
+            range =
+                TextRange::new(range.start() - overall_sub, range.end() - overall_sub - end_sub);
+        }
+        Some(range)
+    }
+}
+
+#[derive(Default)]
+pub struct PositionTransform {
+    pub insertions: Vec<(TextSize, PrettifyWsKind)>,
+}
+
+fn ws_kind_width(ws: PrettifyWsKind) -> u32 {
+    match ws {
+        PrettifyWsKind::Space => 1,
+        PrettifyWsKind::Indent(indent) => 4 * (indent as u32),
+        PrettifyWsKind::Newline => 1,
+    }
```

Comment:
I messed up my commits so the whitespace fixup mapping stuff didnt get its own commit unfortunately but this is the relevant part (super in efficient as one can tell as I hacked it together real quick). There is also an off by one error here somewhere as some offsets seem to not map cleanly causing things to break